### PR TITLE
ARROW-4198: [Gandiva] Added support to cast timestamp

### PR DESCRIPTION
**Pull request by shyambits2004**
Added howard hinnant date project as a third party library.
Used system timezone database for timezone information.

### CODE REVIEW

**Comment by shyambits2004** [comment]
@wesm 

There are less than 10 files in that project and we need the most of them (if I remember it right, I tried including one by one) for time zone support. So, it seemed good idea to add it as a third party library instead of selected files.

**Comment by wesm** [comment]
Right, what I'm saying is that we should vendor everything (that is needed, which may be everything) or nothing

**Comment by shyambits2004** [comment]
A question for my clarity

Why not third party tool chain ? Build time ?

**Comment by wesm** [comment]
Even though it's a small component, we would need to add it to conda-forge or other package managers. It would need to be configured for offline builds (https://github.com/apache/arrow/tree/master/cpp/thirdparty). It is more work than vendoring given that the component is quite simple. Others (@pitrou @xhochy etc) may have other opinions so am interested to see them

**Comment by pitrou** [comment]
I'd like to know which files are used exactly. It seems the PR misses some source file(s) (`cast_time.cc`?).

**Comment by pitrou** [comment]
I would suggest we vendor `tz.h` and `tz.cpp` in addition to `date.h`.

**Comment by wesm** [comment]
Those depend on a couple compatibility headers but overall it seems a light requirement

**Comment by shyambits2004** [comment]
Closed by mistake.

Is there a specific structure for vendored code? Or dump everything into arrow/vendored ?

I am thinking of creating directory structure as src/vendored as a peer to arrow, gandiva and create a sub-folder datetime which will act as place holder for the above files.

Please let me know if you have other suggestions.

**Comment by wesm** [comment]
You can create a subdirectory for this in `arrow/vendored` (since there are multiple files, arrow/vendored/date.h can be nested inside a new directory). Since this change introduces compiled symbols, they may need to be exported with ARROW_EXPORT. There will doubtlessly be other parts of the Arrow project where date/time/tz utilities will be used so having them part of the Arrow tree is the simplest thing right now

I feel that having gandiva as a child of the arrow/ directory (and `arrow::gandiva` namespace) makes as much or more sense than the current structure

https://issues.apache.org/jira/browse/ARROW-3359

**Comment by shyambits2004** [comment]
Made the changes to vendor timezone apis.

I need help in understanding on how ARROW_EXPORT would help in this context. Windows platform?

**Comment by shyambits2004** [comment]
And one more question, Can the code part of vendor directory eliminated from code coverage? We might not use all apis (code) from the vendor.

File: cpp/src/gandiva/precompiled/time.cc [hunk 1]
```diff
@@ -563,4 +563,108 @@ date64 castDATE_utf8(int64_t context, const char* input, int32 length) {
...
+ * ts_fields[2] = day
+ * ts_fields[3] = hours
+ * ts_fields[4] = minutes
+ * ts_fields[5] = seconds
+ * ts_fields[6] = sub_seconds
+ * ts_fields[7] = displacement_hours
+ * ts_fields[8] = displacement_minutes
+ */
+timestamp castTIMESTAMP_utf8(int64_t context, const char* input, int32 length) {
+  int ts_fields[9] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
```
Comment (pravindra) [inline cpp/src/gandiva/precompiled/time.cc]: can we please use enums for the fields ?

File: cpp/src/gandiva/cast_time.cc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/vendored/datetime/tz.h"
+
+extern "C" {
+
+// TODO : Do input validation or make sure the callers do that ?
+int gdv_fn_time_with_zone(int* time_fields, const char* zone, int zone_len,
```
Comment (pravindra) [inline cpp/src/gandiva/cast_time.cc]: pls use enums.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 3]
```diff
@@ -108,6 +108,16 @@ void ExportedStubFunctions::AddMappings(Engine* engine) const {
       "gdv_fn_to_date_utf8_utf8_int32", types->i64_type() /*return_type*/, args,
       reinterpret_cast<void*>(gdv_fn_to_date_utf8_utf8_int32));
 
+  // gdv_fn_time_with_zone
```
Comment (pravindra) [inline cpp/src/gandiva/gdv_function_stubs.cc]: can this be moved to cast_time.cc ?

File: cpp/src/gandiva/precompiled/time.cc (hunk 1 above)
Comment (shyambits2004) [reply cpp/src/gandiva/precompiled/time.cc]: done

File: cpp/src/gandiva/cast_time.cc (hunk 2 above)
Comment (shyambits2004) [reply cpp/src/gandiva/cast_time.cc]: done

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 3 above)
Comment (shyambits2004) [reply cpp/src/gandiva/gdv_function_stubs.cc]: done

File: cpp/src/arrow/util/parsing.h [hunk 4]
```diff
@@ -361,6 +361,12 @@ class StringConverter<Int32Type> : public StringToSignedIntConverterMixin<Int32T
 template <>
 class StringConverter<Int64Type> : public StringToSignedIntConverterMixin<Int64Type> {};
 
+using arrow::util::date::day;
```
Comment (pitrou) [inline cpp/src/arrow/util/parsing.h]: Generally, top-level `using` declarations should only be for `.cc` files. Putting such a declaration in a header file will affect all files including that header files.
(also, in this case it risks produding name collisions, as you've seen below with `year`)

**Review by pitrou — CHANGES_REQUESTED** [review]
Since we're moving `date.h` to a subdirectory and vendoring additional code, you should:
* update the top-level `LICENSE.txt`
* add a `README.md` in `cpp/src/arrow/vendored/datetime` explaining where the vendored files come from (and which revision, probably)

File: cpp/src/arrow/util/parsing.h (hunk 4 above)
Comment (shyambits2004) [reply cpp/src/arrow/util/parsing.h]: Done

**Comment by shyambits2004** [comment]
Made the changes suggested by Antoine.

Travis is failing while building api documentation. Fired travis twice from the morning.

"~/build/apache/arrow/cpp/apidoc ~/build/apache/arrow/python ~/build/apache/arrow
/home/travis/build/apache/arrow/cpp/src/arrow/compute/kernel.h:170: error: The following parameters of arrow::compute::UnaryKernel::Call(FunctionContext *ctx, const Datum &input, Datum *out)=0 are not documented:
  parameter 'ctx'
  parameter 'input' (warning treated as error, aborting now)"

This was passing before and I checked that there are other pull requests whose travis job is failing with this error intermittently.

**Comment by pitrou** [comment]
@shyambits2004 Sorry for that. The doc building failure is fixed on git master, you should rebase.

**Comment by shyambits2004** [comment]
Just a gentle reminder on the review/push of the changes.

**Review by pravindra — APPROVED** [review]

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you @shyambits2004 and @pravindra !

**Pull request closed without merging**